src/frontend/src/binder/query.rs

```diff
@@ -330,6 +334,9 @@ impl Binder {
                     .into());
                 }
 
+                // align the schema
+                let schema = self.bind_set_expr(body)?.schema().clone();
```

Comment:
also, I think it's okay utilizing `bind_set_expr` to align the schema for the current union here - so that we could use this `schema` for subsequent contexts.

cc @chenzl25.